File: lib/cfr/builtins/libs/dart/core/uri.dart
import 'dart:convert';
import 'dart:core';

import 'package:hydro_sdk/cfr/runtimeSupport.dart';

class VMManagedUri extends VMManagedBox<Uri> {
  VMManagedUri(
      {required this.table, required this.vmObject, required this.hydroState})
      : super(
          table: table,
          vmObject: vmObject,
          hydroState: hydroState,
        ) {
    table['getScheme'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.scheme,
      ];
    });
    table['getAuthority'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.authority,
      ];
    });
    table['getUserInfo'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.userInfo,
      ];
    });
    table['getHost'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.host,
      ];
    });
    table['getPort'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.port,
      ];
    });
    table['getPath'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.path,
      ];
    });
    table['getQuery'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.query,
      ];
    });
    table['getFragment'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.fragment,
      ];
    });
    table['getPathSegments'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<List<dynamic>>(
            object: vmObject.pathSegments,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getQueryParameters'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Map>(
            object: vmObject.queryParameters,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getQueryParametersAll'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Map>(
            object: vmObject.queryParametersAll,
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['getIsAbsolute'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isAbsolute,
      ];
    });
    table['getHasScheme'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasScheme,
      ];
    });
    table['getHasAuthority'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasAuthority,
      ];
    });
    table['getHasPort'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasPort,
      ];
    });
    table['getHasQuery'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasQuery,
      ];
    });
    table['getHasFragment'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasFragment,
      ];
    });
    table['getHasEmptyPath'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasEmptyPath,
      ];
    });
    table['getHasAbsolutePath'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hasAbsolutePath,
      ];
    });
    table['getOrigin'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.origin,
      ];
    });
    table['isScheme'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.isScheme(luaCallerArguments[1]),
      ];
    });
    table['toFilePath'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toFilePath(windows: luaCallerArguments[1]['windows']),
      ];
    });
    table['getData'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      final returnValue = vmObject.data;
      if (returnValue != null) {
        return [
          maybeBoxObject<UriData?>(
              object: returnValue, hydroState: hydroState, table: HydroTable()),
        ];
      }
      return [];
    });
    table['getHashCode'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.hashCode,
      ];
    });
    table['toString'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        vmObject.toString(),
      ];
    });
    table['replace'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uri>(
            object: vmObject.replace(
                fragment: luaCallerArguments[1]['fragment'],
                host: luaCallerArguments[1]['host'],
                path: luaCallerArguments[1]['path'],
                pathSegments: maybeUnBoxAndBuildArgument<Iterable<String>?>(
                    luaCallerArguments[1]['pathSegments'],
                    parentState: hydroState),
                port: luaCallerArguments[1]['port'],
                query: luaCallerArguments[1]['query'],
                queryParameters:
                    maybeUnBoxAndBuildArgument<Map<String, dynamic>?>(
                        luaCallerArguments[1]['queryParameters'],
                        parentState: hydroState),
                scheme: luaCallerArguments[1]['scheme'],
                userInfo: luaCallerArguments[1]['userInfo']),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['removeFragment'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uri>(
            object: vmObject.removeFragment(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['resolve'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uri>(
            object: vmObject.resolve(luaCallerArguments[1]),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['resolveUri'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uri>(
            object: vmObject.resolveUri(maybeUnBoxAndBuildArgument<Uri>(
                luaCallerArguments[1],
                parentState: hydroState)),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
    table['normalizePath'] =
        makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
      return [
        maybeBoxObject<Uri>(
            object: vmObject.normalizePath(),
            hydroState: hydroState,
            table: HydroTable()),
      ];
    });
  }

  final HydroTable table;

  final HydroState hydroState;

  final Uri vmObject;
}

void loadUri({required HydroState hydroState, required HydroTable table}) {
  table['uri'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri(
              fragment: luaCallerArguments[1]['fragment'],
              host: luaCallerArguments[1]['host'],
              path: luaCallerArguments[1]['path'],
              pathSegments: maybeUnBoxAndBuildArgument<Iterable<String>?>(
                  luaCallerArguments[1]['pathSegments'],
                  parentState: hydroState),
              port: luaCallerArguments[1]['port'],
              query: luaCallerArguments[1]['query'],
              queryParameters:
                  maybeUnBoxAndBuildArgument<Map<String, dynamic>?>(
                      luaCallerArguments[1]['queryParameters'],
                      parentState: hydroState),
              scheme: luaCallerArguments[1]['scheme'],
              userInfo: luaCallerArguments[1]['userInfo']),
          hydroState: hydroState,
          table: luaCallerArguments[0])
    ];
  });
  table['uriHttp'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.http(
              luaCallerArguments[1],
              luaCallerArguments[2],
              maybeUnBoxAndBuildArgument<Map<String, dynamic>?>(
                  luaCallerArguments[3],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriHttps'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.https(
              luaCallerArguments[1],
              luaCallerArguments[2],
              maybeUnBoxAndBuildArgument<Map<String, dynamic>?>(
                  luaCallerArguments[3],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriFile'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.file(luaCallerArguments[1],
              windows: luaCallerArguments[2]['windows']),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriDirectory'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.directory(luaCallerArguments[1],
              windows: luaCallerArguments[2]['windows']),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriDataFromString'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.dataFromString(luaCallerArguments[1],
              base64: luaCallerArguments[2]['base64'],
              encoding: maybeUnBoxAndBuildArgument<Encoding?>(
                  luaCallerArguments[2]['encoding'],
                  parentState: hydroState),
              mimeType: luaCallerArguments[2]['mimeType'],
              parameters: maybeUnBoxAndBuildArgument<Map<String, String>?>(
                  luaCallerArguments[2]['parameters'],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriDataFromBytes'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.dataFromBytes(
              maybeUnBoxAndBuildArgument<List<int>>(luaCallerArguments[1],
                  parentState: hydroState),
              mimeType: luaCallerArguments[2]['mimeType'],
              parameters: maybeUnBoxAndBuildArgument<Map<String, String>?>(
                  luaCallerArguments[2]['parameters'],
                  parentState: hydroState),
              percentEncoded: luaCallerArguments[2]['percentEncoded']),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriBase'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.base, hydroState: hydroState, table: HydroTable()),
    ];
  });
  table['uriParse'] = makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Uri>(
          object: Uri.parse(luaCallerArguments[1], luaCallerArguments[2],
              luaCallerArguments[3]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriTryParse'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    final returnValue = Uri.tryParse(
        luaCallerArguments[1], luaCallerArguments[2], luaCallerArguments[3]);
    if (returnValue != null) {
      return [
        maybeBoxObject<Uri?>(
            object: returnValue, hydroState: hydroState, table: HydroTable()),
      ];
    }
    return [];
  });
  table['uriEncodeComponent'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.encodeComponent(luaCallerArguments[1]),
    ];
  });
  table['uriEncodeQueryComponent'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.encodeQueryComponent(luaCallerArguments[1],
          encoding: maybeUnBoxAndBuildArgument<Encoding>(
              luaCallerArguments[2]['encoding'],
              parentState: hydroState)),
    ];
  });
  table['uriDecodeComponent'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.decodeComponent(luaCallerArguments[1]),
    ];
  });
  table['uriDecodeQueryComponent'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.decodeQueryComponent(luaCallerArguments[1],
          encoding: maybeUnBoxAndBuildArgument<Encoding>(
              luaCallerArguments[2]['encoding'],
              parentState: hydroState)),
    ];
  });
  table['uriEncodeFull'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.encodeFull(luaCallerArguments[1]),
    ];
  });
  table['uriDecodeFull'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      Uri.decodeFull(luaCallerArguments[1]),
    ];
  });
  table['uriSplitQueryString'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<Map>(
          object: Uri.splitQueryString(luaCallerArguments[1],
              encoding: maybeUnBoxAndBuildArgument<Encoding>(
                  luaCallerArguments[2]['encoding'],
                  parentState: hydroState)),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriParseIPv4Address'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<List<dynamic>>(
          object: Uri.parseIPv4Address(luaCallerArguments[1]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  table['uriParseIPv6Address'] =
      makeLuaDartFunc(func: (List<dynamic> luaCallerArguments) {
    return [
      maybeBoxObject<List<dynamic>>(
          object: Uri.parseIPv6Address(luaCallerArguments[1],
              luaCallerArguments[2], luaCallerArguments[3]),
          hydroState: hydroState,
          table: HydroTable()),
    ];
  });
  registerBoxer<Uri>(boxer: (
      {required Uri vmObject,
      required HydroState hydroState,
      required HydroTable table}) {
    return VMManagedUri(
        vmObject: vmObject, hydroState: hydroState, table: table);
  });
}

File: runtime/dart/core/uri.ts
import { IEncoding } from "../convert/encoding";
import { IIterable } from "./iterable";
import { IList } from "./list";
import { IUriData } from "./uriData";
declare const dart: {
    core: {
        uri: (
            this: void,
            uri: IUri,
            props: {
                fragment?: string | undefined;
                host?: string | undefined;
                path?: string | undefined;
                pathSegments?: IIterable<string> | undefined;
                port?: number | undefined;
                query?: string | undefined;
                queryParameters?: { [index: string]: any } | undefined;
                scheme?: string | undefined;
                userInfo?: string | undefined;
            }
        ) => IUri;
        uriHttp: (
            authority: string,
            unencodedPath: string,
            queryParameters?: { [index: string]: any } | undefined
        ) => IUri;
        uriHttps: (
            authority: string,
            unencodedPath: string,
            queryParameters?: { [index: string]: any } | undefined
        ) => IUri;
        uriFile: (
            path: string,
            props: { windows?: boolean | undefined }
        ) => IUri;
        uriDirectory: (
            path: string,
            props: { windows?: boolean | undefined }
        ) => IUri;
        uriDataFromString: (
            content: string,
            props: {
                base64: boolean;
                encoding?: IEncoding | undefined;
                mimeType?: string | undefined;
                parameters?: { [index: string]: string } | undefined;
            }
        ) => IUri;
        uriDataFromBytes: (
            bytes: IList<number>,
            props: {
                mimeType: string;
                parameters?: { [index: string]: string } | undefined;
                percentEncoded: boolean;
            }
        ) => IUri;
        uriBase: () => IUri;
        uriParse: (
            uri: string,
            start: number,
            end?: number | undefined
        ) => IUri;
        uriTryParse: (
            uri: string,
            start: number,
            end?: number | undefined
        ) => IUri | undefined;
        uriEncodeComponent: (component: string) => string;
        uriEncodeQueryComponent: (
            component: string,
            props: { encoding: IEncoding }
        ) => string;
        uriDecodeComponent: (encodedComponent: string) => string;
        uriDecodeQueryComponent: (
            encodedComponent: string,
            props: { encoding: IEncoding }
        ) => string;
        uriEncodeFull: (uri: string) => string;
        uriDecodeFull: (uri: string) => string;
        uriSplitQueryString: (
            query: string,
            props: { encoding: IEncoding }
        ) => { [index: string]: string };
        uriParseIPv4Address: (host: string) => IList<number>;
        uriParseIPv6Address: (
            host: string,
            start: number,
            end?: number | undefined
        ) => IList<number>;
    };
};
export interface IUri {
    getScheme: () => string;
    getAuthority: () => string;
    getUserInfo: () => string;
    getHost: () => string;
    getPort: () => number;
    getPath: () => string;
    getQuery: () => string;
    getFragment: () => string;
    getPathSegments: () => IList<string>;
    getQueryParameters: () => { [index: string]: string };
    getQueryParametersAll: () => { [index: string]: IList<string> };
    getIsAbsolute: () => boolean;
    getHasScheme: () => boolean;
    getHasAuthority: () => boolean;
    getHasPort: () => boolean;
    getHasQuery: () => boolean;
    getHasFragment: () => boolean;
    getHasEmptyPath: () => boolean;
    getHasAbsolutePath: () => boolean;
    getOrigin: () => string;
    isScheme: (scheme: string) => boolean;
    toFilePath: (props: { windows?: boolean | undefined }) => string;
    getData: () => IUriData | undefined;
    getHashCode: () => number;
    toString: () => string;
    replace: (props: {
        fragment?: string | undefined;
        host?: string | undefined;
        path?: string | undefined;
        pathSegments?: IIterable<string> | undefined;
        port?: number | undefined;
        query?: string | undefined;
        queryParameters?: { [index: string]: any } | undefined;
        scheme?: string | undefined;
        userInfo?: string | undefined;
    }) => IUri;
    removeFragment: () => IUri;
    resolve: (reference: string) => IUri;
    resolveUri: (reference: IUri) => IUri;
    normalizePath: () => IUri;
}
export class Uri {
    public constructor(props: {
        fragment?: string | undefined;
        host?: string | undefined;
        path?: string | undefined;
        pathSegments?: IIterable<string> | undefined;
        port?: number | undefined;
        query?: string | undefined;
        queryParameters?: { [index: string]: any } | undefined;
        scheme?: string | undefined;
        userInfo?: string | undefined;
    }) {
        dart.core.uri(this, props);
    }
    public static http(
        authority: string,
        unencodedPath: string,
        queryParameters?: { [index: string]: any } | undefined
    ): IUri {
        return dart.core.uriHttp(authority, unencodedPath, queryParameters);
    }
    public static https(
        authority: string,
        unencodedPath: string,
        queryParameters?: { [index: string]: any } | undefined
    ): IUri {
        return dart.core.uriHttps(authority, unencodedPath, queryParameters);
    }
    public static file(
        path: string,
        props: { windows?: boolean | undefined }
    ): IUri {
        return dart.core.uriFile(path, props);
    }
    public static directory(
        path: string,
        props: { windows?: boolean | undefined }
    ): IUri {
        return dart.core.uriDirectory(path, props);
    }
    public static dataFromString(
        content: string,
        props: {
            base64?: boolean;
            encoding?: IEncoding | undefined;
            mimeType?: string | undefined;
            parameters?: { [index: string]: string } | undefined;
        }
    ): IUri {
        return dart.core.uriDataFromString(content, {
            ...dataFromStringDefaultProps,
            ...props,
        });
    }
    public static dataFromBytes(
        bytes: IList<number>,
        props: {
            mimeType?: string;
            parameters?: { [index: string]: string } | undefined;
            percentEncoded?: boolean;
        }
    ): IUri {
        return dart.core.uriDataFromBytes(bytes, {
            ...dataFromBytesDefaultProps,
            ...props,
        });
    }
    public static base(): IUri {
        return dart.core.uriBase();
    }
    public static parse(
        uri: string,
        start: number = 0,
        end?: number | undefined
    ): IUri {
        return dart.core.uriParse(uri, start, end);
    }
    public static tryParse(
        uri: string,
        start: number = 0,
        end?: number | undefined
    ): IUri | undefined {
        return dart.core.uriTryParse(uri, start, end);
    }
    public static encodeComponent(component: string): string {
        return dart.core.uriEncodeComponent(component);
    }
    public static encodeQueryComponent(
        component: string,
        props: { encoding?: IEncoding }
    ): string {
        return dart.core.uriEncodeQueryComponent(component, {
            ...encodeQueryComponentDefaultProps,
            ...props,
        });
    }
    public static decodeComponent(encodedComponent: string): string {
        return dart.core.uriDecodeComponent(encodedComponent);
    }
    public static decodeQueryComponent(
        encodedComponent: string,
        props: { encoding?: IEncoding }
    ): string {
        return dart.core.uriDecodeQueryComponent(encodedComponent, {
            ...decodeQueryComponentDefaultProps,
            ...props,
        });
    }
    public static encodeFull(uri: string): string {
        return dart.core.uriEncodeFull(uri);
    }
    public static decodeFull(uri: string): string {
        return dart.core.uriDecodeFull(uri);
    }
    public static splitQueryString(
        query: string,
        props: { encoding?: IEncoding }
    ): { [index: string]: string } {
        return dart.core.uriSplitQueryString(query, {
            ...splitQueryStringDefaultProps,
            ...props,
        });
    }
    public static parseIPv4Address(host: string): IList<number> {
        return dart.core.uriParseIPv4Address(host);
    }
    public static parseIPv6Address(
        host: string,
        start: number = 0,
        end?: number | undefined
    ): IList<number> {
        return dart.core.uriParseIPv6Address(host, start, end);
    }
    private readonly _dart_getScheme: () => string = undefined as any;
    private readonly _dart_getAuthority: () => string = undefined as any;
    private readonly _dart_getUserInfo: () => string = undefined as any;
    private readonly _dart_getHost: () => string = undefined as any;
    private readonly _dart_getPort: () => number = undefined as any;
    private readonly _dart_getPath: () => string = undefined as any;
    private readonly _dart_getQuery: () => string = undefined as any;
    private readonly _dart_getFragment: () => string = undefined as any;
    private readonly _dart_getPathSegments: () => IList<string> =
        undefined as any;
    private readonly _dart_getQueryParameters: () => {
        [index: string]: string;
    } = undefined as any;
    private readonly _dart_getQueryParametersAll: () => {
        [index: string]: IList<string>;
    } = undefined as any;
    private readonly _dart_getIsAbsolute: () => boolean = undefined as any;
    private readonly _dart_getHasScheme: () => boolean = undefined as any;
    private readonly _dart_getHasAuthority: () => boolean = undefined as any;
    private readonly _dart_getHasPort: () => boolean = undefined as any;
    private readonly _dart_getHasQuery: () => boolean = undefined as any;
    private readonly _dart_getHasFragment: () => boolean = undefined as any;
    private readonly _dart_getHasEmptyPath: () => boolean = undefined as any;
    private readonly _dart_getHasAbsolutePath: () => boolean = undefined as any;
    private readonly _dart_getOrigin: () => string = undefined as any;
    private readonly _dart_isScheme: (scheme: string) => boolean =
        undefined as any;
    private readonly _dart_toFilePath: (props: {
        windows?: boolean | undefined;
    }) => string = undefined as any;
    private readonly _dart_getData: () => IUriData | undefined =
        undefined as any;
    private readonly _dart_getHashCode: () => number = undefined as any;
    private readonly _dart_toString: () => string = undefined as any;
    private readonly _dart_replace: (props: {
        fragment?: string | undefined;
        host?: string | undefined;
        path?: string | undefined;
        pathSegments?: IIterable<string> | undefined;
        port?: number | undefined;
        query?: string | undefined;
        queryParameters?: { [index: string]: any } | undefined;
        scheme?: string | undefined;
        userInfo?: string | undefined;
    }) => IUri = undefined as any;
    private readonly _dart_removeFragment: () => IUri = undefined as any;
    private readonly _dart_resolve: (reference: string) => IUri =
        undefined as any;
    private readonly _dart_resolveUri: (reference: IUri) => IUri =
        undefined as any;
    private readonly _dart_normalizePath: () => IUri = undefined as any;
    public getScheme(): string {
        return this._dart_getScheme();
    }
    public getAuthority(): string {
        return this._dart_getAuthority();
    }
    public getUserInfo(): string {
        return this._dart_getUserInfo();
    }
    public getHost(): string {
        return this._dart_getHost();
    }
    public getPort(): number {
        return this._dart_getPort();
    }
    public getPath(): string {
        return this._dart_getPath();
    }
    public getQuery(): string {
        return this._dart_getQuery();
    }
    public getFragment(): string {
        return this._dart_getFragment();
    }
    public getPathSegments(): IList<string> {
        return this._dart_getPathSegments();
    }
    public getQueryParameters(): { [index: string]: string } {
        return this._dart_getQueryParameters();
    }
    public getQueryParametersAll(): { [index: string]: IList<string> } {
        return this._dart_getQueryParametersAll();
    }
    public getIsAbsolute(): boolean {
        return this._dart_getIsAbsolute();
    }
    public getHasScheme(): boolean {
        return this._dart_getHasScheme();
    }
    public getHasAuthority(): boolean {
        return this._dart_getHasAuthority();
    }
    public getHasPort(): boolean {
        return this._dart_getHasPort();
    }
    public getHasQuery(): boolean {
        return this._dart_getHasQuery();
    }
    public getHasFragment(): boolean {
        return this._dart_getHasFragment();
    }
    public getHasEmptyPath(): boolean {
        return this._dart_getHasEmptyPath();
    }
    public getHasAbsolutePath(): boolean {
        return this._dart_getHasAbsolutePath();
    }
    public getOrigin(): string {
        return this._dart_getOrigin();
    }
    public isScheme(scheme: string): boolean {
        return this._dart_isScheme(scheme);
    }
    public toFilePath(props: { windows?: boolean | undefined }): string {
        return this._dart_toFilePath(props);
    }
    public getData(): IUriData | undefined {
        return this._dart_getData();
    }
    public getHashCode(): number {
        return this._dart_getHashCode();
    }
    public toString(): string {
        return this._dart_toString();
    }
    public replace(props: {
        fragment?: string | undefined;
        host?: string | undefined;
        path?: string | undefined;
        pathSegments?: IIterable<string> | undefined;
        port?: number | undefined;
        query?: string | undefined;
        queryParameters?: { [index: string]: any } | undefined;
        scheme?: string | undefined;
        userInfo?: string | undefined;
    }): IUri {
        return this._dart_replace(props);
    }
    public removeFragment(): IUri {
        return this._dart_removeFragment();
    }
    public resolve(reference: string): IUri {
        return this._dart_resolve(reference);
    }
    public resolveUri(reference: IUri): IUri {
        return this._dart_resolveUri(reference);
    }
    public normalizePath(): IUri {
        return this._dart_normalizePath();
    }
}
const dataFromStringDefaultProps = {
    base64: false,
};
const dataFromBytesDefaultProps = {
    mimeType: "application/octet-stream",
    percentEncoded: false,
};
const encodeQueryComponentDefaultProps = {
    encoding: utf8,
};
const decodeQueryComponentDefaultProps = {
    encoding: utf8,
};
const splitQueryStringDefaultProps = {
    encoding: utf8,
};
